plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/CostingFun/Core.hs

```diff
@@ -249,61 +257,85 @@ runCostingFunTwoArguments
     => CostingFun ModelTwoArguments
     -> a1
     -> a2
-    -> ExBudget
+    -> ExBudgetStream
 runCostingFunTwoArguments (CostingFun cpu mem) =
     case (runTwoArgumentModel cpu, runTwoArgumentModel mem) of
         (!runCpu, !runMem) -> onMemoryUsages $ \mem1 mem2 ->
-            ExBudget (ExCPU    $ runCpu mem1 mem2)
-                     (ExMemory $ runMem mem1 mem2)
+            zipCostStream
+                (runCpu mem1 mem2)
+                (runMem mem1 mem2)
 {-# INLINE runCostingFunTwoArguments #-}
 
 runTwoArgumentModel
     :: ModelTwoArguments
-    -> ExMemory
-    -> ExMemory
-    -> CostingInteger
+    -> CostStream
+    -> CostStream
+    -> CostStream
+runTwoArgumentModel
+    (ModelTwoArgumentsConstantCost c) = lazy $ \_ _ -> CostLast c
 runTwoArgumentModel
-    (ModelTwoArgumentsConstantCost c) = lazy $ \_ _ -> c
+    (ModelTwoArgumentsAddedSizes (ModelAddedSizes intercept slope)) =
+        lazy $ \costs1 costs2 ->
+            scaleLinearly intercept slope $ addCostStream costs1 costs2
 runTwoArgumentModel
-    (ModelTwoArgumentsAddedSizes (ModelAddedSizes intercept slope)) = \(ExMemory size1) (ExMemory size2) ->
-        (size1 + size2) * slope + intercept -- TODO is this even correct? If not, adjust the other implementations too.
+    (ModelTwoArgumentsSubtractedSizes (ModelSubtractedSizes intercept slope minSize)) =
+        lazy $ \costs1 costs2 -> do
+            let !size1 = sumCostStream costs1
+                !size2 = sumCostStream costs2
```

Comment:
If we wanted to, we could've opted for convertnig `costs*` to `size*` automatically via a type class, but I thought it wouldn't be worth it.

Reply:
Yeah, we shouldn't make the slowest thing the easiest.